implement use cases for creating, editing, deleting, and fetching project states, and define StateRepository interface

ibrahimdeeb0 opened the pull request.

ibrahimdeeb0 committed:
feature(state): implement use cases for creating, editing, deleting, and fetching project states, and define StateRepository interface

ibrahimdeeb0 committed:
Merge remote-tracking branch 'origin/development' into feature/project-state-crud

# Conflicts:
#	src/main/kotlin/model/Project.kt
#	src/main/kotlin/model/State.kt

ibrahimdeeb0 committed:
test(state): add unit tests for Create, Edit, Delete, and GetAllStates use cases

ibrahimdeeb0 committed:
resolve conflicts from merging development into feature/project-state-crud

Farah315 commented on src/main/kotlin/Main.kt:
```diff
@@ -1,7 +1,5 @@
 
-
-
 fun main() {
```
Main function is empty

Farah315 commented on src/main/kotlin/Main.kt:
```diff
@@ -1,7 +1,5 @@
 
-
-
 fun main() {
```
Main function is empty. You should avoid including files with no actual changes

ibrahimdeeb0 replied:
Just for the format purpose, I deleted the two empty lines

Farah315 commented on src/main/kotlin/logic/exceptions/UnauthorizedAccessException.kt:
```diff
@@ -0,0 +1,10 @@
+    message: String = NOT_AUTHORIZED_EXCEPTION_MESSAGE,
+) : Exception(message) {
+
+    companion object {
+        const val NOT_AUTHORIZED_EXCEPTION_MESSAGE = "User is not authorized to perform this action."
```
I think it’s better to place it at the end of the class where the use case is used, just like in the previous project
and Exceptions not exceptions

ibrahimdeeb0 replied:
What if the same exception message should appear in multiple files?
should avoid duplication and be more readable, yes?

Farah315 replied:
If the same exception message appears in multiple files, consider extracting it into a constant to avoid duplication and improve readability

Farah315 replied:
+ Change the file name and make it Exceptions

Farah315 replied:
Also, this file must be in model

Farah315 commented on src/main/kotlin/logic/usecases/state/DeleteStateUseCase.kt:
```diff
@@ -0,0 +1,25 @@
+        )
+    }
+
+    companion object {
```
As long as you write it here, there is no need for that object in the Exceptions class

ibrahimdeeb0 replied:
Sorry, I will modify this to be like CreateStateUseCase

ibrahimdeeb0 replied:
I modified it

ibrahimdeeb0 committed:
replace generic exception with UnauthorizedAccessException in use cases

Farah315 commented on src/test/kotlin/logic/usecases/state/CreateStateUseCaseTest.kt:
```diff
@@ -0,0 +1,41 @@
+
+    private lateinit var stateRepository: FakeStateRepository
+    private lateinit var createStateUseCase: CreateStateUseCase
+    private val projectId = "P001"
```
The project ID is in UUID format (e.g., `761b124d-e7a4-4957-ba9f-692b330fee09`). Since this is fake data, it’s probably fine for now — but keep in mind this format when validating or generating real project IDs later

ibrahimdeeb0 replied:
will not affect in these cases, but I will keep it in my mind

hazemka reviewed (dismissed):
- replace state as string with State model in each function
- continue with implementation after merge

ibrahimdeeb0 committed:
update use cases to match StateRepository using State object not stateName as a String

ibrahimdeeb0 committed:
Fix: Fix parameter order bug in EditStateUseCase and finalize State model handling
- Cleaned up FakeStateRepository to use consistent ID-based deletion
- All test cases now pass successfully

ibrahimdeeb0 committed:
fix conflict

ibrahimdeeb0 committed:
fix issues in development branch

ibrahimdeeb0 committed:
move exceptions package from logic layer to model layer

ibrahimdeeb0 committed:
fix Issues on TaskRepositoryImpl and make it abstract class
